File: code/src/main/java/it/unifi/nave/data/Hashable.java
package it.unifi.nave.data;

import it.unifi.nave.crypto.CryptoFactory;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.ObjectOutputStream;
import java.io.Serializable;

public interface Hashable extends Serializable {
  default String hash() {
    return CryptoFactory.newHashUtil().hash(serialize());
  }

  default byte[] serialize() {
    try {
      ByteArrayOutputStream arrayOutputStream = new ByteArrayOutputStream();
      ObjectOutputStream objectOutputStream = new ObjectOutputStream(arrayOutputStream);
      objectOutputStream.writeObject(this);
      return arrayOutputStream.toByteArray();
    } catch (IOException e) {
      e.printStackTrace();
      throw new RuntimeException("Something went wrong", e);
    }
  }
}

File: UniBlock/src/main/scala/data/Hashable.scala
package it.unifi.nave.uniblock
package data

import it.unifi.nave.uniblock.helper.crypto.HashHelper

import java.io.{ByteArrayOutputStream, ObjectOutputStream}

trait Hashable extends Serializable {
  def hash: String = HashHelper.hash(Left(serialize))

  def serialize: Array[Byte] = {
    val arrayOutputStream = new ByteArrayOutputStream
    val objectOutputStream = new ObjectOutputStream(arrayOutputStream)
    objectOutputStream.writeObject(this)
    arrayOutputStream.toByteArray
  }
}
